allow an sql suffix to be added to the resulting INSERT SQL

darinpp opened the pull request:
MySQL allows an option ON DUPLICATE KEY UPDATE suffix for INSERT statements. This PR adds an ability to add the logic as string when doing the bulk insert.
```
sql_suffix = ' ON DUPLICATE KEY UPDATE count=count+1'
Book.bulk_insert(sql_suffix: sql_suffix) do |worker|
  book_attrs.each do |attrs|
    worker.add(attrs)
  end
end
```

jamis commented:
Thanks for the pull request! However, I don't know that I like plugging raw SQL into the query like that. I think I'd rather see it as part of the API, perhaps exposed like this:

```ruby
Book.bulk_insert(on_duplicate_key: { update: 'count=count+1' }) do
```

or maybe on the worker directly:

```ruby
worker.on_duplicate_key update: 'count=count+1'
```

Or both. Or maybe something else entirely. :) Also, if you update this PR, would also add a test or two to demonstrate that your changes work as expected?

Thanks!

darinpp commented:
Ok, changed it to your first suggestion. Added a test as well. Your tests
use SQLite so it isn't possible to actually test the result of the SQL
execution.
Thanks,
Darin

On Wed, Feb 15, 2017 at 11:56 AM, Jamis Buck <notifications@github.com>
wrote:

> Thanks for the pull request! However, I don't know that I like plugging
> raw SQL into the query like that. I think I'd rather see it as part of the
> API, perhaps exposed like this:
>
> Book.bulk_insert(on_duplicate_key: { update: 'count=count+1' }) do
>
> or maybe on the worker directly:
>
> worker.on_duplicate_key update: 'count=count+1'
>
> Or both. Or maybe something else entirely. :) Also, if you update this PR,
> would also add a test or two to demonstrate that your changes work as
> expected?
>
> Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jamis/bulk_insert/pull/18#issuecomment-280120836>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGudlXwcbxeVvLco-TEjXmJPmPoiZsqkks5rc1h1gaJpZM4MBLiT>
> .
>

jamis commented:
@darinpp, thanks for your quick turn around on my suggestion, and I'm sorry for taking so long to get back to you. I like how this looks, but there's another PR (#21) that is proposing a refactoring of the query builder code. Would you be okay waiting until that PR lands? Then we can rework this as an extension of those changes, and everything should play nicely.

jamis commented on test/bulk_insert_test.rb:
```diff
@@ -34,5 +34,18 @@ class BulkInsertTest < ActiveSupport::TestCase
+  test "accepts on_duplicate_key argument and constructs sql_on_duplicate in the worker" do
+    worker = Testing.bulk_insert(on_duplicate_key: { update: '123'})
+
+    assert_equal worker.instance_eval { @sql_on_duplicate }, ' on duplicate key update 123'
```
The more I think about this, the more I feel uncomfortable about peeking at an object's internal state. Rather than reaching inside the object and violating encapsulation, I'd like to see another way to test this. I know SQLite doesn't support the feature, but maybe there's another way to do this. PR #21 is proposing a refactoring that would expose a method that returns the generated SQL. Maybe this test could query that SQL to make sure that the expected snippet exists? You'll need to wait until that PR lands, but I think it would expose a better way to implement this test.

darinpp commented:
I don't see how the `before_save_callback` gives access to the sql. However - I see your point. Take a look at how I changed the test. It stubs the `execute` of the connection and asserts that the correct sql has been generated.

jamis commented:
I'm so sorry, I linked to the wrong pull request--I meant to point you at #19, specifically where it is pulling the query generation into another method. With access to that method, you could then test the SQL directly, without having to peek inside and without mocking. The stub you've got now is definitely cleaner than before, but I think it could be even cleaner with the refactored method to produce the SQL.

Thanks for giving some thought to my suggestion, and finding a way to get rid of the "peeking" at the instance variable. Let me think about this for a bit. I think maybe it's "good enough" as it is, but it's a bit late here and I'm not thinking clearly. :)

jamis commented:
@darinpp -- sorry to sit on this so long. PR #19 has landed in master finally--would you please fix the conflicts on this branch? Then we can focus on getting this one merged into master, too. Thank you!

jamis commented:
As there hasn't been any further work on this pull request, I'm going to go ahead and close it. Anyone wishing to work on it further may feel free to reopen at that time.
